proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+  "setReplyTCB param_a param_b \<lbrace>\<lambda>s. P (list_refs_of_replies' s)\<rbrace>"
+  apply (wpsimp simp: setReplyTCB_def)
+  apply (erule arg_cong[where f=P, THEN iffD1, rotated])
+  apply (rule ext)
+  apply (clarsimp simp: opt_map_def map_set_def list_refs_of_reply'_def obj_at'_def projectKO_eq)
+  done
+
+lemma setObject_reply_pde_mappings'[wp]:
+  "setObject p (r::reply) \<lbrace>valid_pde_mappings'\<rbrace>"
+  by (wp valid_pde_mappings_lift')
+
+crunches replyPush
+  for pspace_aligned'[wp]: pspace_aligned'
+  and pspace_distinct'[wp]: pspace_distinct'
+  and if_unsafe_then_cap'[wp]: "if_unsafe_then_cap'"
+  and valid_global_refs'[wp]: "valid_global_refs'"
+  and valid_arch_state'[wp]: "valid_arch_state'"
+  and valid_irq_node'[wp]: "\<lambda>s. valid_irq_node' (irq_node' s) s"
+  and valid_irq_handlers'[wp]: "valid_irq_handlers'"
+  and valid_irq_states'[wp]: "valid_irq_states'"
+  and valid_machine_state'[wp]: "valid_machine_state'"
+  and irqs_masked'[wp]: "irqs_masked'"
+  and valid_release_queue'[wp]: "valid_release_queue'"
+  and ct_not_inQ[wp]: "ct_not_inQ"
+  and ct_idle_or_in_cur_domain'[wp]: "ct_idle_or_in_cur_domain'"
+  and valid_pde_mappings'[wp]: "valid_pde_mappings'"
+  and pspace_domain_valid[wp]: "pspace_domain_valid"
+  and ksCurDomain[wp]: "\<lambda>s. P (ksCurDomain s)"
+  and valid_dom_schedule'[wp]: "valid_dom_schedule'"
+  and cur_tcb'[wp]: "cur_tcb'"
+  and no_0_obj'[wp]: no_0_obj'
+  and valid_mdb'[wp]: valid_mdb'
+  and tcb_at'[wp]: "tcb_at' t"
+  and cte_wp_at'[wp]: "cte_wp_at' P p"
+  and ctes_of[wp]: "\<lambda>s. P (ctes_of s)"
+  (wp: crunch_wps hoare_vcg_all_lift valid_irq_node_lift simp: crunch_simps valid_mdb'_def)
+
+lemma tcbSchedEnqueue_valid_tcb_state'[wp]:
+  "tcbSchedEnqueue t \<lbrace>valid_tcb_state' ts\<rbrace>"
+  unfolding tcbSchedEnqueue_def
+  apply (wp, rule threadSet_wp)
+        apply (rule_tac Q="\<lambda>_. valid_tcb_state' ts" in hoare_strengthen_post[rotated])
+         apply (fastforce simp: valid_tcb_state'_def valid_bound_obj'_def obj_at'_def projectKO_eq
+                                projectKO_ep projectKO_reply projectKO_ntfn projectKO_tcb
+                         split: thread_state.splits option.splits)
+        apply wpsimp+
+  done
+
+crunches rescheduleRequired, scheduleTCB
+  for valid_tcb_state'[wp]: "valid_tcb_state' ts"
+  (wp: crunch_wps)
+
+lemma setThreadState_valid_tcb_state'[wp]:
+  "setThreadState st p \<lbrace>valid_tcb_state' ts\<rbrace>"
+  by (wpsimp wp: threadSet_wp simp: setThreadState_def)
+     (fastforce simp: valid_tcb_state'_def valid_bound_obj'_def obj_at'_def projectKO_eq
+                      projectKO_ep projectKO_reply projectKO_ntfn projectKO_tcb
+               split: thread_state.splits option.splits)
+
+crunches setThreadState
+  for sc_at'_n[wp]: "sc_at'_n n p"
+  and sc_ko_at'[wp]: "\<lambda>s. P (ko_at' (sc :: sched_context) p s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+definition replyPush' where
```

Comment:
Maybe it'll become evident as I keep reviewing, but could you give me a little explanation as to why you want this new definition?

Reply:
Some wp rules require valid_objs' as a precondition to schedContextDonate. Having a monadic rewrite rule for replyPush means we only need to prove that valid_objs' is invariant over replyPush'.